proto/automation.proto

```diff
@@ -0,0 +1,91 @@
+// Protobuf messages for the automation framework.
+
+// Messages (e.g. Task) are used both for checkpoint data and API access
+// (e.g. retrieving the current status of a pending cluster operation).
+
+syntax = "proto3";
+
+package automation;
+
+enum ClusterOperationState {
+  UNKNOWN_CLUSTER_OPERATION_STATE = 0;
+  CLUSTER_OPERATION_NOT_STARTED = 1;
+  CLUSTER_OPERATION_RUNNING = 2;
+  CLUSTER_OPERATION_DONE = 3;
+}
+
+message ClusterOperation {
+  string id = 1;
+  // TaskContainer are processed sequentially, one at a time.
+  repeated TaskContainer serial_tasks = 2;
+  // Cached value. This has to be re-evaluated e.g. after a checkpoint load because running tasks may have already finished.
+  ClusterOperationState state = 3;
+  // Error of the first task which failed. Set after state advanced to CLUSTER_OPERATION_DONE. If empty, all tasks succeeded. Cached value, see state above.
+  string error = 4;
+}
+
+// TaskContainer holds one or more task which may be executed in parallel.
+// "concurrency", if > 0, limits the amount of concurrently executed tasks.
+// TODO(mberlin): Find a better name. "TaskGroup"?
+message TaskContainer {
+  repeated Task parallel_tasks = 1;
+  int32 concurrency = 2;
+}
+
+enum TaskState {
+  UNKNOWN_TASK_STATE = 0;
+  NOT_STARTED = 1;
+  RUNNING = 2;
+  DONE = 3;
+}
+
+// Task represents a specific task which should be automatically executed.
+message Task {
+  // Task specification.
+  string name = 1;
+  map <string, string> parameters = 2;
+
+  // Runtime data.
+  string id = 3;
+  TaskState state = 4;
+  // Set after state advanced to DONE. If empty, the task did succeed.
```

Comment:
DONE

This was meant for "error", but then I added "output" later. I moved it to "error".